File: src/Butil/Bit.Butil/Publics/StorageBuckets.cs
using System;
using System.Diagnostics.CodeAnalysis;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps the <see href="https://developer.mozilla.org/en-US/docs/Web/API/Storage_Buckets_API">Storage Buckets API</see>
/// (<c>navigator.storageBuckets</c>): named compartments of the origin's storage, each with its own
/// quota, persistence, durability and expiry.
/// </summary>
/// <remarks>
/// Without buckets an origin has one pile of storage that the browser evicts whole: the draft the
/// user has not sent yet goes out with the thumbnail cache. A bucket is the unit that fixes that -
/// put the re-fetchable things in one bucket and the irreplaceable ones in another, and only the
/// first is worth evicting.
/// <br/>
/// Each bucket owns its own IndexedDB databases, Cache Storage caches and file system. The file
/// members here are that file system - the same
/// <see cref="OriginPrivateFileSystem">origin private file system</see>, rooted in the bucket rather
/// than in the origin.
/// <br/>
/// Chromium-only at the time of writing. Every member degrades to a false/null/empty answer where
/// the API is absent, so a call is safe to make before checking <see cref="IsSupported"/> - it just
/// won't do anything.
/// </remarks>
[ButilService(typeof(StorageBuckets))]
public class StorageBuckets(IJSRuntime js)
{
    /// <summary>True when the runtime exposes <c>navigator.storageBuckets</c>.</summary>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.storageBuckets.isSupported");

    /// <summary>
    /// Opens the named bucket, creating it when it isn't there.
    /// </summary>
    /// <param name="name">
    /// The bucket name: lower-case letters, digits, <c>-</c> and <c>_</c>, not starting with <c>-</c>
    /// or <c>_</c>. A name outside that alphabet is rejected by the browser and comes back as null.
    /// </param>
    /// <param name="persisted">Ask for a bucket exempt from eviction. The browser decides - read <see cref="StorageBucketInfo.Persisted"/> for the answer.</param>
    /// <param name="durability">How hard the browser should try not to lose the latest writes.</param>
    /// <param name="quota">A ceiling in bytes for this bucket, which can only be lower than the origin's own quota.</param>
    /// <param name="expires">When the browser may delete the bucket. Null leaves it without an expiry.</param>
    /// <returns>The bucket's state, or null when the API is absent or the name was rejected.</returns>
    /// <remarks>
    /// The options only apply when the bucket is created: opening an existing bucket returns it as
    /// it is, so this is safe to call on every start-up.
    /// </remarks>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(StorageBucketInfo))]
    public ValueTask<StorageBucketInfo?> Open(string name,
                                              bool persisted = false,
                                              StorageBucketDurability durability = StorageBucketDurability.Default,
                                              long? quota = null,
                                              DateTimeOffset? expires = null)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(name);
        return js.Invoke<StorageBucketInfo?>("BitButil.storageBuckets.open",
            name, persisted, Durability(durability), quota, expires?.ToUnixTimeMilliseconds());
    }

    /// <summary>Reads a bucket's current state without changing any of it.</summary>
    /// <param name="name">The bucket name.</param>
    /// <returns>The bucket's state, or null when the API is absent.</returns>
    /// <remarks>
    /// Opening a bucket is what creates it, so asking about a name that was never used creates an
    /// empty bucket under that name. Check <see cref="Keys"/> first when that matters.
    /// </remarks>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(StorageBucketInfo))]
    public ValueTask<StorageBucketInfo?> Get(string name)
        => js.Invoke<StorageBucketInfo?>("BitButil.storageBuckets.get", name);

    /// <summary>Lists the names of every bucket this origin has, in alphabetical order.</summary>
    public ValueTask<string[]> Keys() => js.Invoke<string[]>("BitButil.storageBuckets.keys");

    /// <summary>
    /// Deletes a bucket and everything in it - its databases, caches and files.
    /// </summary>
    /// <param name="name">The bucket name.</param>
    /// <returns>True when the call succeeded. Deleting a bucket that isn't there is not an error.</returns>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> Delete(string name) => js.Invoke<bool>("BitButil.storageBuckets.delete", name);

    /// <summary>Asks the browser to make this bucket persistent. The user agent decides whether to grant.</summary>
    /// <param name="name">The bucket name.</param>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> Persist(string name) => js.Invoke<bool>("BitButil.storageBuckets.persist", name);

    /// <summary>True when this bucket is exempt from eviction.</summary>
    /// <param name="name">The bucket name.</param>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> Persisted(string name) => js.Invoke<bool>("BitButil.storageBuckets.persisted", name);

    /// <summary>
    /// This bucket's own quota and usage, in bytes.
    /// </summary>
    /// <param name="name">The bucket name.</param>
    /// <remarks>
    /// <see cref="StorageEstimate.UsageDetails"/> is always empty here - a bucket reports one number
    /// for itself, with no per-API breakdown.
    /// </remarks>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(StorageEstimate))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(StorageUsageDetail))]
    public ValueTask<StorageEstimate> Estimate(string name)
        => js.Invoke<StorageEstimate>("BitButil.storageBuckets.estimate", name);

    /// <summary>
    /// Sets the point after which the browser may delete the bucket.
    /// </summary>
    /// <param name="name">The bucket name.</param>
    /// <param name="expires">When the bucket becomes disposable.</param>
    /// <remarks>
    /// An expiry is permission to delete, not a scheduled deletion: the data usually survives past
    /// it, and reading it back afterwards is not something to rely on.
    /// <br/>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> SetExpires(string name, DateTimeOffset expires)
        => js.Invoke<bool>("BitButil.storageBuckets.setExpires", name, expires.ToUnixTimeMilliseconds());

    /// <summary>Reads the bucket's expiry, or null when it has none.</summary>
    /// <param name="name">The bucket name.</param>
    public async ValueTask<DateTimeOffset?> GetExpires(string name)
    {
        var expires = await js.Invoke<long?>("BitButil.storageBuckets.getExpires", name);
        return expires is null ? null : DateTimeOffset.FromUnixTimeMilliseconds(expires.Value);
    }

    /// <summary>
    /// Lists a directory inside the bucket's own file system. Not recursive.
    /// </summary>
    /// <param name="name">The bucket name.</param>
    /// <param name="path">A directory path inside the bucket, or empty for its root.</param>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(OpfsEntry))]
    public ValueTask<OpfsEntry[]> List(string name, string path = "")
        => js.Invoke<OpfsEntry[]>("BitButil.storageBuckets.list", name, path);

    /// <summary>Reads a file inside the bucket as text.</summary>
    /// <param name="name">The bucket name.</param>
    /// <param name="path">A file path inside the bucket.</param>
    /// <returns>The text, or null when there is no file there.</returns>
    public ValueTask<string?> ReadText(string name, string path)
        => js.Invoke<string?>("BitButil.storageBuckets.readText", name, path);

    /// <summary>Writes text to a file inside the bucket, creating what's missing along the path.</summary>
    /// <param name="name">The bucket name.</param>
    /// <param name="path">A file path inside the bucket.</param>
    /// <param name="text">The contents to write.</param>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> WriteText(string name, string path, string text)
        => js.Invoke<bool>("BitButil.storageBuckets.write", name, path, text, null);

    /// <summary>Writes bytes to a file inside the bucket, creating what's missing along the path.</summary>
    /// <param name="name">The bucket name.</param>
    /// <param name="path">A file path inside the bucket.</param>
    /// <param name="data">The contents to write.</param>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> WriteBytes(string name, string path, byte[] data)
        => js.Invoke<bool>("BitButil.storageBuckets.write", name, path, null, data);

    /// <summary>Deletes a file or directory inside the bucket.</summary>
    /// <param name="name">The bucket name.</param>
    /// <param name="path">The path inside the bucket to delete.</param>
    /// <param name="recursive">Required to delete a directory that isn't empty.</param>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> Remove(string name, string path, bool recursive = false)
        => js.Invoke<bool>("BitButil.storageBuckets.remove", name, path, recursive);

    private static string? Durability(StorageBucketDurability durability) => durability switch
    {
        StorageBucketDurability.Relaxed => "relaxed",
        StorageBucketDurability.Strict => "strict",
        _ => null,
    };
}

File: src/Butil/Bit.Butil/Scripts/storageBuckets.ts
var BitButil = (window as any).BitButil = (window as any).BitButil || {};

(function (butil: any) {
    // Buckets are addressed by name on every call rather than kept in a handle registry: open() is
    // idempotent - it returns the existing bucket when one already exists under that name - so a
    // registry would only add a way for .NET and the browser to disagree about what is still there.
    // The cost of that is that open() also *creates*, which is why only open() and write() reach
    // for a bucket directly; everything else goes through existing().

    function manager() { return (window.navigator as any).storageBuckets; }

    function open(name: string, options: any) {
        const buckets = manager();
        if (!buckets?.open) return Promise.reject(new Error('Storage Buckets are not supported'));
        return buckets.open(name, options ?? undefined);
    }

    // open() is create-or-get, so reaching for a bucket by name is what brings it into existence.
    // Everything that only makes sense against a bucket that is already there goes through this
    // instead, so asking about an unknown name answers "no" rather than quietly creating it.
    async function existing(name: string) {
        const buckets = manager();
        if (!buckets?.keys) throw new Error('Storage Buckets are not supported');
        const names = await buckets.keys();
        if (!names.includes(name)) throw new Error(`there is no bucket named '${name}'`);
        return await open(name, undefined);
    }

    async function describe(bucket: any) {
        // Every field is its own promise on the bucket, and each is optional across engines - one
        // that is missing must not lose the rest, hence the individual guards.
        const persisted = bucket.persisted ? await bucket.persisted() : false;
        const durability = bucket.durability ? await bucket.durability() : null;
        const expires = bucket.expires ? await bucket.expires() : null;
        const estimate = bucket.estimate ? await bucket.estimate() : null;

        return {
            name: bucket.name ?? '',
            persisted: !!persisted,
            durability: durability ?? null,
            expires: typeof expires === 'number' ? expires : null,
            quota: typeof estimate?.quota === 'number' ? estimate.quota : null,
            usage: typeof estimate?.usage === 'number' ? estimate.usage : null
        };
    }

    butil.storageBuckets = {
        isSupported() { return !!(window.navigator as any).storageBuckets; },
        async open(name: string, persisted: boolean, durability: string | null, quota: number | null, expires: number | null) {
            const options: any = {};
            if (persisted) options.persisted = true;
            if (durability) options.durability = durability;
            if (quota !== null && quota !== undefined) options.quota = quota;
            if (expires !== null && expires !== undefined) options.expires = expires;

            try { return await describe(await open(name, options)); }
            catch { return null; }
        },
        async keys() {
            const buckets = manager();
            if (!buckets?.keys) return [];
            try { return await buckets.keys(); } catch { return []; }
        },
        async delete(name: string) {
            const buckets = manager();
            if (!buckets?.delete) return false;
            try { await buckets.delete(name); return true; } catch { return false; }
        },
        async get(name: string) {
            try { return await describe(await existing(name)); } catch { return null; }
        },
        async persist(name: string) {
            try {
                const bucket = await existing(name);
                return bucket.persist ? await bucket.persist() : false;
            } catch {
                return false;
            }
        },
        async persisted(name: string) {
            try {
                const bucket = await existing(name);
                return bucket.persisted ? await bucket.persisted() : false;
            } catch {
                return false;
            }
        },
        async estimate(name: string) {
            try {
                const bucket = await existing(name);
                const estimate = bucket.estimate ? await bucket.estimate() : null;
                return {
                    quota: typeof estimate?.quota === 'number' ? estimate.quota : null,
                    usage: typeof estimate?.usage === 'number' ? estimate.usage : null,
                    // A bucket reports one number for the whole bucket; the per-API breakdown of
                    // navigator.storage.estimate() has no equivalent here.
                    usageDetails: []
                };
            } catch {
                return { quota: null, usage: null, usageDetails: [] };
            }
        },
        async setExpires(name: string, expires: number) {
            try {
                const bucket = await existing(name);
                if (!bucket.setExpires) return false;
                await bucket.setExpires(expires);
                return true;
            } catch {
                return false;
            }
        },
        async getExpires(name: string) {
            try {
                const bucket = await existing(name);
                const expires = bucket.expires ? await bucket.expires() : null;
                return typeof expires === 'number' ? expires : null;
            } catch {
                return null;
            }
        },

        // A bucket owns a whole OPFS tree of its own, reached the same way as the origin's - so the
        // file operations are butil.originPrivateFileSystem's, run against the bucket's root.
        async list(name: string, path: string) {
            try { return await butil.originPrivateFileSystem.listUnder(await bucketRoot(name, false), path); }
            catch { return []; }
        },
        async readText(name: string, path: string) {
            try { return await butil.originPrivateFileSystem.readTextUnder(await bucketRoot(name, false), path); }
            catch { return null; }
        },
        async write(name: string, path: string, text: string | null, bytes: Uint8Array | null) {
            // The one file member that creates: a write names what it wants to exist, so the bucket
            // is created along with the directories on the path when it isn't there yet.
            try { return await butil.originPrivateFileSystem.writeUnder(await bucketRoot(name, true), path, text, bytes); }
            catch { return false; }
        },
        async remove(name: string, path: string, recursive: boolean) {
            try { return await butil.originPrivateFileSystem.removeUnder(await bucketRoot(name, false), path, recursive); }
            catch { return false; }
        }
    };

    async function bucketRoot(name: string, create: boolean) {
        const bucket = create ? await open(name, undefined) : await existing(name);
        if (!bucket.getDirectory) throw new Error('this bucket exposes no file system');
        return await bucket.getDirectory();
    }
}(BitButil));
